Libraries/LibGUI/TextEditor.cpp

```diff
@@ -1460,6 +1462,9 @@ void TextEditor::for_each_visual_line(size_t line_index, Callback callback) cons
     size_t start_of_line = 0;
     size_t visual_line_index = 0;
 
+    if (!is_visual_data_up_to_date())
+        return;
```

Comment:
This doesn't feel like a good fix.
Instead of making `for_each_visual_line()` a no-op while visual line data is invalid, we should be checking for this at the call site instead.
Who is trying to iterate over the visual lines while they are invalid and why? And how can we fix it? :)

Reply:
I was looking into this, and it is what is causing #2498. When typing a letter to replace a selection, the TextDocument first performs a RemoveTextCommand, which causes the TextDocumentLine's codepoints vector to get nulled out (note: only when the entire line is being replaced). This causes the call to `content_x_for_position` when trying to set the cursor to its new position to fail because it directly accesses the Vector's backing data, which is still null.

Reply:
@awesomekling as @FalseHonesty this happens through `content_x_for_position`. I added the check there, which is still not ideal, but it seems to do the job.